File: scripts/python/leerSerial.py
import serial
import json # Importamos la biblioteca JSON

# Reemplaza 'TU_PUERTO_COM' con tu puerto (ej. 'COM3' o '/dev/tty.usbmodem14202')
PUERTO_SERIAL = 'COM3'
BAUD_RATE = 38400

try:
    # Configura y abre la conexión serial
    ser = serial.Serial(PUERTO_SERIAL, BAUD_RATE, timeout=1)
    print(f"Conectado a {PUERTO_SERIAL} a {BAUD_RATE} baudios.")
    print("Presiona Ctrl+C para detener.")

    while True:
        # Lee una línea completa (hasta el salto de línea '\n')
        linea_leida = ser.readline()

        # Si la línea no está vacía...
        if linea_leida:
            # 1. Decodifica de bytes a string (utf-8) y quita espacios en blanco
            json_string = linea_leida.decode('utf-8').strip()

            # 2. Intenta parsear (convertir) el string JSON a un objeto Python
            try:
                datos = json.loads(json_string)
                print(f"Datos recibidos: {datos}")
                # Puedes acceder a los datos individualmente:
                # print(f"Temperatura: {datos['temp_c']} °C")

            except json.JSONDecodeError:
                # Esto pasa si el string no es un JSON válido (ej. un mensaje a medias)
                print(f"Error: No se pudo decodificar JSON. Recibido: {json_string}")

except serial.SerialException as e:
    print(f"Error al abrir el puerto serial: {e}")
except KeyboardInterrupt:
    print("\nLectura detenida por el usuario.")
finally:
    if 'ser' in locals() and ser.is_open:
        ser.close()
        print("Conexión serial cerrada.")

File: scripts/go/leerSerial.go
package main

import (
	"bufio"    // Para leer líneas fácilmente
	"encoding/json" // Para parsear JSON
	"fmt"
	"log"
	"time"

	"github.com/tarm/serial"
)

// Definimos una struct que coincida con nuestro JSON
// Las etiquetas `json:"..."` le dicen a Go cómo mapear los campos
type DatosMicrobit struct {
	Temperatura float64 `json:"temp_c"`
	Luz         int     `json:"luz"`
	AccelX      int     `json:"accel_x"`
}

func main() {
	// Reemplaza 'TU_PUERTO_COM' con tu puerto
	PUERTO := "COM3"
	BAUD_RATE := 38400

	// 1. Configurar el puerto serie
	config := &serial.Config{
		Name:        PUERTO,
		Baud:        BAUD_RATE,
		ReadTimeout: time.Second * 1, // Timeout de lectura
	}

	// 2. Abrir el puerto
	stream, err := serial.OpenPort(config)
	if err != nil {
		log.Fatalf("Error al abrir puerto serial: %v", err)
	}
	// Aseguramos que el puerto se cierre al final
	defer stream.Close()

	log.Printf("Conectado a %s a %d baudios.\n", PUERTO, BAUD_RATE)

	// 3. Crear un "reader" de buffer para leer línea por línea
	reader := bufio.NewReader(stream)

	for {
		// 4. Leer hasta el próximo salto de línea '\n'
		jsonString, err := reader.ReadString('\n')
		if err != nil {
			log.Printf("Error leyendo del serial: %v", err)
			continue // Intenta de nuevo
		}

		if jsonString == "" {
			continue // Línea vacía
		}

		// 5. Parsear el JSON.
		// Creamos una variable del tipo de nuestra struct
		var datos DatosMicrobit

		// Usamos Unmarshal para convertir el string (en bytes) a la struct
		err = json.Unmarshal([]byte(jsonString), &datos)
		if err != nil {
			log.Printf("Error al parsear JSON: %v. Recibido: %s", err, jsonString)
			continue
		}

		// 6. ¡Éxito! Imprimir los datos
		fmt.Printf("Datos recibidos: %+v\n", datos)
		// fmt.Printf("Temperatura: %.2f °C\n", datos.Temperatura)
	}
}